### ISSUE TITLE
3.1.0-rc Exception - EntityChangesToAuditEntryConverter::findOrCreateAuditEntry "More than one result was found for query although one row or none was expected"

### ISSUE THREAD

**Issue by fyrye** [oroinc/platform#882]
**Environment:**
oroinc/crm-application 3.1.0-RC (new install)
PHP 7.2.13

**Issue:**
When the message consumers are running, they error constantly at this point, preventing the message consumer from continuing. This results in the log file on production that grows very quickly. In an 8 hour span the log file was 2GB. 

**Notes:**
I ran the ` oro:message-queue:consume --env=prod -vv` from the terminal to determine the source of the issue.

It appears that there is a `version` unique constraint on the `Audit` entity, that is not being taken into consideration in the query. 

```php
$auditEntry = $this->getAuditRepository($auditEntryClass)
            ->createQueryBuilder('a')
            ->select('a')
            ->where('a.transactionId = :transactionId AND a.objectClass = :objectClass AND a.objectId = :objectId')
            ->setParameter('transactionId', $transactionId)
            ->setParameter('objectClass', $entityClass)
            ->setParameter('objectId', $entityId)
            ->getQuery()
            ->getOneOrNullResult();
```

A similar issue #725 reference a similar issue with the unique constraint failing upon update.

**Comment by Yuri47h** [comment]
I have the same issue.

**Comment by aivus** [comment]
Hi.

Thank you guys for reporting this issue.
Internal ticket has been created: BAP-18992

**Comment by Yuri47h** [comment]
@aivus Hi. 
What status this issue? Where you planing fix it? Because we have problems with magento integration through this bug.

**Comment by anyt** [comment]
It's not planned to be fixed soon. Feel free to contribute.

**Comment by fyrye** [comment]
A quick and dirty fix would be to add `->setMaxResults(1)->orderBy('a.version', 'DESC')` to the query builder. However, I am not eligible to contribute and am unsure if it would be the most desirable approach.

```php
$auditEntry = $this->getAuditRepository($auditEntryClass)
            ->createQueryBuilder('a')
            ->select('a')
            ->where('a.transactionId = :transactionId AND a.objectClass = :objectClass AND a.objectId = :objectId')
            ->setMaxResuls(1)
            ->orderBy('a.version', 'DESC')
            ->setParameter('transactionId', $transactionId)
            ->setParameter('objectClass', $entityClass)
            ->setParameter('objectId', $entityId)
            ->getQuery()
            ->getOneOrNullResult();
```

**Comment by dxops** [comment]
The reason is that \Oro\Bundle\DataAuditBundle\Service\EntityChangesToAuditEntryConverter::findOrCreateAuditEntry will not work with async processing when two different consumers create similar audit entries between getOneOrNullResult and flush calls. Should be UPSERT query or try-catch block to catch duplicate record exception on insert and reload it.

**Comment by Yuri47h** [comment]
@dxops Consumer create similar audit entries only when one message from oro_message_queue will processing two different consumers. Is it possible?

**Comment by dxops** [comment]
No, consumer will not process messages twice. Even redelivered message will work because it find will locale proper entry. I might be wrong with my async reason, because I can't find obvious place to duplicate or share the same transaction_id

**Comment by Yuri47h** [comment]
Thank`s, i realized. I try fix it.

**Comment by Yuri47h** [comment]
I found the source of the problem. 
In `DataAuditBundle/Async/AuditChangedEntitiesProcessor.php`

we have 
```
        if ($body['collections_updated']) {
            $this->messageProducer->send(Topics::ENTITIES_RELATIONS_CHANGED, $nextMessage);
        }
        $this->messageProducer->send(Topics::ENTITIES_INVERSED_RELATIONS_CHANGED, $nextMessage);
```
Еhe same message is sent to different topics and if you have many consumers for example - 4 -  different consumers process this two different topics with the same message at the same time -`findOneOrNull` not found entity because not yet created and inserted to DB two oro_audit for one transaction.

I see two way to fix:
1. Send only one message and merge `AuditChangedEntitiesInverseRelationsProcessor` and `AuditChangedEntitiesRelationsProcessor` in one processor
2. First send message from `AuditChangedEntitiesProcessor` to `AuditChangedEntitiesInverseRelationsProcessor` and next - send message from `AuditChangedEntitiesInverseRelationsProcessor` to `AuditChangedEntitiesRelationsProcessor`

**Comment by sadortun** [comment]
Thanks @mario-opinionsystem for your fix

**Comment by sadortun** [comment]
Also, should we have an UNIQUE key on [ transaction_id, object_id , object_class ] ? That would prevent the duplicate insertion.

**Comment by aivus** [comment]
Hello.

This issue has been fixed in 3.1.21 and 4.1.2.

**Comment by anyt** [comment]
The issue was fixed in OroPlatform CE 4.1.2, and OroPlatform Enterprise 3.1.22
